Integrate VideoPress with Divi Builder

pgk opened the pull request:
#### Changes proposed in this Pull Request:

<!--- Explain what functional changes your PR includes -->
* Adds very basic Divi Builder support.
* You can embed by guid or by video url.
* The video will grow to fit the container.
* No support for video privacy at this point. Videos need to be public for this to work.

Note: Originally made using [create-divi-extension](https://github.com/elegantthemes/create-divi-extension). Later the code was adapted to fit in the VideoPress/Jetpack build system and conventions.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

pxWta-1KB-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Run install and build for the plugin you are using, e.g. `jetpack install plugins/jetpack` and `jetpack build plugins/jetpack`. Note: You may need to build dependencies, at least for the first time. You can do that with e.g. `jetpack build --all`
* You will need to install Divi. If you are an Automattician and need Divi, refer to pxWta-1KB-p2 
* Create a new post and use the Divi builder.
* Create a layout and then search for "videopress". Select that.
* Test pasting both guids and urls.
* You can see a VideoPress embed

pgk committed:
[not verified] Videopress Divi Extension.

pgk committed:
[not verified] Changelog entry.

pgk committed:
[not verified] Match render method members to parent.

roundhill commented on projects/packages/videopress/src/videopress-divi/class-videopress-divi-module.php:
```diff
@@ -0,0 +1,92 @@
+	public function get_fields() {
+		return array(
+			'guid' => array(
+				'label'           => esc_html__( 'Guid or URL', 'jetpack-videopress-pkg' ),
```
How about `URL or Video ID`?

roundhill commented on projects/packages/videopress/src/videopress-divi/class-videopress-divi-module.php:
```diff
@@ -0,0 +1,92 @@
+				'label'           => esc_html__( 'Guid or URL', 'jetpack-videopress-pkg' ),
+				'type'            => 'text',
+				'option_category' => 'basic_option',
+				'description'     => esc_html__( 'paste a url or guid', 'jetpack-videopress-pkg' ),
```
`Paste a URL or Video ID?`

roundhill commented on projects/packages/videopress/src/videopress-divi/class-videopress-divi-module.php:
```diff
@@ -0,0 +1,92 @@
+			esc_html( $this->props['guid'] )
+		);
+		$iframe_src = sprintf(
+			'https://videopress.com/embed/%s?hd=0&autoPlay=0&permalink=0&loop=0',
```
I think the `hd` parameter isn't necessary

roundhill reviewed (commented):
Haven't tested it completely through yet but have some copy thoughts

pgk committed:
[not verified] Fix error when guid does not match.

pgk committed:
[not verified] Copy changes.

pgk committed:
Resize VideoPress iframe to fit container.

pgk committed:
Fix linter errors

pgk committed:
Merge remote-tracking branch 'origin/trunk' into add/videopress-divi

pgk committed:
Add embedder query param.

pgk committed:
Add tracks.

pgk committed:
Merge remote-tracking branch 'origin/trunk' into add/videopress-divi

roundhill reviewed (changes requested):
I'm still getting this error as soon as I click on the VideoPress block (I've built the Jetpack plugin):

<img width="1031" alt="Screenshot 2023-02-06 at 11 07 55 AM" src="https://user-images.githubusercontent.com/789137/217051430-2c68a7c6-7d8c-4de2-97fd-0e565fb74e8c.png">

And is it possible to use a different icon? Maybe the VP logo?

<img width="185" alt="Screenshot 2023-02-06 at 11 10 04 AM" src="https://user-images.githubusercontent.com/789137/217051502-88061f13-f50b-4f10-9042-f2a207d94f8c.png">

pgk committed:
MOre compatible with jetpack.

roundhill commented on projects/packages/videopress/src/class-initializer.php:
```diff
@@ -126,6 +126,9 @@ private static function active_initialization() {
 			Admin_UI::init();
 		}
+
+		/* require_once __DIR__ . '/class-divi.php'; */
```
Remove?

roundhill committed:
Set the icon to "q" which is the icon font for the video icon.

roundhill commented:
I snuck in a change to use the same video icon for our block: 

<img width="186" alt="Screenshot 2023-02-10 at 11 03 16 AM" src="https://user-images.githubusercontent.com/789137/218164132-39b112cf-1032-426b-bfef-936fcf27972c.png">

roundhill commented on projects/packages/videopress/changelog/add-videopress-divi:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: added
+
+Adds Divi Builder Compatibility.
```
```suggestion
Adds Divi Builder Compatibility for VideoPress.
```

roundhill reviewed (dismissed):
This is working well for me now. I've tested it with and without Divi installed to ensure it is being loaded properly. Resizing with `videopress-iframe.js` is working well too.

pgk committed:
Update projects/packages/videopress/changelog/add-videopress-divi

Co-authored-by: Daniel Roundhill <dan@automattic.com>

pgk committed:
Rename method that initializes the extension.

pgk committed:
Remove commented code.

roundhill reviewed (dismissed):
Latest changes look good

zinigor commented on 3:
```diff
@@ -0,0 +1,78 @@
+<?php
```
Is this really a file called `3`? Or is there some kind of problem with this diff?

pgk replied:
removed!

pgk committed:
Merge remote-tracking branch 'origin/trunk' into add/videopress-divi

pgk committed:
Delete weird file.

pgk committed:
Merge remote-tracking branch 'origin/trunk' into add/videopress-divi

pgk committed:
Run tools/fixup-project-versions.sh

pgk committed:
Remove package.json created from create-divi-extension.

pgk committed:
Merge remote-tracking branch 'origin/trunk' into add/videopress-divi

pgk committed:
Rename mixed case folder to lowercase.

pgk committed:
Merge remote-tracking branch 'origin/trunk' into add/videopress-divi

pgk committed:
Ignore phpcs complaining about not using variables that are part of a
method signature.

pgk committed:
Update changelogs.

pgk committed:
Ignore enqueued script lint error.

pgk committed:
Again, ignore linting error we can't fix because of inheritance.

pgk committed:
Eslint fixes.

jeherve commented on projects/packages/videopress/src/class-divi.php:
```diff
@@ -0,0 +1,74 @@
+	 * Creates the extension's main class instance.
+	 */
+	public function initialize_extension() {
+		require_once plugin_dir_path( __FILE__ ) . 'videopress-divi/class-videopress-divi-extension.php';
```
Do we need to require the file, given that we're in the package where the classes should be autoloaded?

pgk replied:
Unfortunately, because it is subclassing a Divi class, we need to ensure Divi is properly loaded, before requiring it:

`add_action( 'divi_extensions_init', array( $this, 'initialize_extension' ) );`

jeherve commented on projects/packages/videopress/src/client/divi-editor/modules/videopress/index.jsx:
```diff
@@ -0,0 +1,41 @@
+		return (
+			<div className="vidi-videopress-wrapper">
+				<iframe title={ iframeTitle } src={ url } width="100%" height="100%"></iframe>
+				<script src="https://en.wordpress.com/wp-content/plugins/video/assets/js/next/videopress-iframe.js?m=1658739239"></script>
```
In other places, we load a file with the same name from `https://s0.wp.com/wp-content/plugins/video/assets/js/next/videopress-iframe.js`, and also from `https://videopress.com/videopress-iframe.js`? Does this matter?

jeherve commented on projects/packages/videopress/src/videopress-divi/README.md:
```diff
@@ -0,0 +1,2 @@
+The VideoPress Divi Extention was initially created using [Create Divi Extension](https://github.com/elegantthemes/create-divi-extension).
```
```suggestion
The VideoPress Divi Extension was initially created using [Create Divi Extension](https://github.com/elegantthemes/create-divi-extension).
```

jeherve commented on projects/packages/videopress/src/videopress-divi/class-videopress-divi-extension.php:
```diff
@@ -0,0 +1,136 @@
+	/**
+	 * The gettext domain for the extension's translations.
+	 *
+	 * @since 1.0.0
```
Will we version the extension differently, or could we use the package number with `$$next-version$$`?

pgk replied:
A subjective matter which I am not feeling strong about. Making the version be the package number sounds ok.

jeherve commented on projects/packages/videopress/src/videopress-divi/class-videopress-divi-extension.php:
```diff
@@ -0,0 +1,136 @@
+	 * @override
+	 */
+	public function hook_et_builder_ready() {
+		require_once plugin_dir_path( __FILE__ ) . 'class-videopress-divi-module.php';
```
Same as before, I was wondering if we needed the require here?

pgk replied:
Unfortunately, this is also subclassing a Divi class

jeherve commented on projects/packages/videopress/src/videopress-divi/class-videopress-divi-extension.php:
```diff
@@ -0,0 +1,136 @@
+	 * Register the extension's js assets.
+	 */
+	public function register_js_assets() {
+		Assets::register_script(
+			self::JETPACK_VIDEOPRESS_DIVI_PKG_NAMESPACE,
+			'../../build/divi-editor/index.js',
+			__FILE__,
+			array(
+				'in_footer'  => true,
+				'css_path'   => '../../build/divi-editor/index.css',
+				'textdomain' => 'jetpack-videopress-pkg',
+			)
+		);
+
+		Assets::enqueue_script( self::JETPACK_VIDEOPRESS_DIVI_PKG_NAMESPACE );
```
We can combine this with the `enqueue` parameter.

```suggestion
		Assets::register_script(
			self::JETPACK_VIDEOPRESS_DIVI_PKG_NAMESPACE,
			'../../build/divi-editor/index.js',
			__FILE__,
			array(
				'in_footer'  => true,
				'css_path'   => '../../build/divi-editor/index.css',
				'textdomain' => 'jetpack-videopress-pkg',
				'enqueue'    => true,
			)
		);
```

However, won't this mean that the file will be enqueued everywhere on the frontend, even when there are no videos?

jeherve commented on projects/packages/videopress/src/videopress-divi/class-videopress-divi-extension.php:
```diff
@@ -0,0 +1,136 @@
+		Assets::enqueue_script( self::JETPACK_VIDEOPRESS_DIVI_PKG_NAMESPACE );
+	}
+
+	/**
+	 * Enqueue frontend stuff.
+	 *
+	 * @override
+	 */
+	public function wp_hook_enqueue_scripts() {
+	}
```
Is is an option to get rid of this entirely, since it does nothing?

jeherve commented on projects/packages/videopress/src/videopress-divi/class-videopress-divi-extension.php:
```diff
@@ -0,0 +1,136 @@
+	 * Register the extension's js assets.
+	 */
+	public function register_js_assets() {
+		Assets::register_script(
```
Do we need to add jQuery as a dependency, since the script seems to rely on it? Or maybe we could get rid of jQuery in `projects/packages/videopress/src/client/divi-editor/index.js`?

pgk replied:
Added!

jeherve commented on projects/packages/videopress/src/client/divi-editor/modules/videopress/index.jsx:
```diff
@@ -0,0 +1,41 @@
+	/^(?:https?:\/\/)?((?:video|video\.word)press\.com\/(?:v|embed)\/)?(?<guid>[a-zA-Z\d]{7,})(?:.*)?/;
+
+class VideoPress extends Component {
+	static slug = 'vidi_videopress';
```
Was it meant to say `divi_videopress`, or is that on purpose to avoid conflicts?

pgk replied:
fixed

jeherve commented on projects/packages/videopress/src/videopress-divi/class-videopress-divi-module.php:
```diff
@@ -0,0 +1,114 @@
+	public function get_fields() {
+		return array(
+			'guid' => array(
+				'label'           => esc_html__( 'URL or Video ID', 'jetpack-videopress-pkg' ),
+				'type'            => 'text',
+				'option_category' => 'basic_option',
+				'description'     => esc_html__( 'Paste a URL or Video ID', 'jetpack-videopress-pkg' ),
```
Would it be worth saying "VideoPress URL" in those strings to avoid confusion?

jeherve commented on projects/packages/videopress/src/videopress-divi/styles/.gitignore:
Do we need this empty file?

pgk replied:
removed

jeherve commented on projects/packages/videopress/src/videopress-divi/scripts/frontend.js:
Do we need this empty file?

pgk replied:
removed

jeherve reviewed (commented):
I gave the code a quick look, left some questions below.

Beyond this, is there by any chance a P2 discussion or something where I could find out more about this project? Maybe a place where I could get my hands on Divi to be able to do some proper testing? :)

If so, could you add the link to the section in the PR description?

Thanks!

pgk committed:
Update projects/packages/videopress/src/videopress-divi/README.md

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

pgk commented:
> Beyond this, is there by any chance a P2 discussion or something where I could find out more about this project?

Updated the description with a link to the PT and added divi-related links there.

pgk committed:
[not verified] Tweak version.

pgk committed:
[not verified] Add jquery as a dep of the Divi extension pack.

pgk committed:
Remove styles folder.

pgk committed:
Tweak translation string.

pgk committed:
Rename js extension slug.

pgk committed:
Remove frontend.js

pgk committed:
[not verified] Merge remote-tracking branch 'origin/trunk' into add/videopress-divi

pgk committed:
Merge remote-tracking branch 'origin/trunk' into add/videopress-divi

pgk committed:
Rename vidi -> divi

pgk committed:
Merge remote-tracking branch 'origin/trunk' into add/videopress-divi

jeherve reviewed (changes requested):
Could you try to fix the failing tests, by updating this PR with latest `trunk`, and then run `tools/fixup-project-versions.sh`?

pgk committed:
Merge remote-tracking branch 'origin/trunk' into add/videopress-divi

pgk committed:
Update monorepo lock

pgk committed:
Update versions.

jeherve reviewed (approved):
This tests well for me. 🚢
